src/kms.rs

```diff
@@ -0,0 +1,329 @@
+    db: DumbBuffer,
+
+    /// Equivalent mapping for reading.
+    zeroes: Box<[u32]>,
+}
+
+impl KmsImpl {
+    /// Create a new KMS backend.
+    ///
+    /// # Safety
+    ///
+    /// The plane must be valid for the lifetime of the backend.
+    pub(crate) unsafe fn new(
+        window_handle: DrmWindowHandle,
+        display: Rc<KmsDisplayImpl>,
+    ) -> Result<Self, SoftBufferError> {
+        log::trace!("new: window_handle={:X}", window_handle.plane);
+
+        // Make sure that the window handle is valid.
+        let plane_handle = match NonZeroU32::new(window_handle.plane) {
+            Some(handle) => plane::Handle::from(handle),
+            None => return Err(SoftBufferError::IncompleteWindowHandle),
+        };
+
+        let plane_info = display
+            .get_plane(plane_handle)
+            .swbuf_err("failed to get plane info")?;
+        let handles = display
+            .resource_handles()
+            .swbuf_err("failed to get resource handles")?;
+
+        // Use either the attached CRTC or the primary CRTC.
+        let crtc = match plane_info.crtc() {
+            Some(crtc) => crtc,
+            None => {
+                log::warn!("no CRTC attached to plane, falling back to primary CRTC");
+                handles
+                    .filter_crtcs(plane_info.possible_crtcs())
+                    .first()
+                    .copied()
+                    .swbuf_err("failed to find a primary CRTC")?
+            }
+        };
+
+        // Use a preferred connector or just select the first one.
+        let connector = handles
+            .connectors
+            .iter()
+            .flat_map(|handle| display.get_connector(*handle, false))
+            .find_map(|conn| {
+                if conn.state() == connector::State::Connected {
+                    Some(conn.handle())
+                } else {
+                    None
+                }
+            })
+            .or_else(|| handles.connectors.first().copied())
+            .swbuf_err("failed to find a valid connector")?;
+
+        Ok(Self {
+            crtc: display
+                .get_crtc(crtc)
+                .swbuf_err("failed to get CRTC info")?,
+            connector,
+            display,
+            buffer: None,
+        })
+    }
+
+    /// Resize the internal buffer to the given size.
+    pub(crate) fn resize(
+        &mut self,
+        width: NonZeroU32,
+        height: NonZeroU32,
+    ) -> Result<(), SoftBufferError> {
+        // Don't resize if we don't have to.
+        if let Some(buffer) = &self.buffer {
+            let (buffer_width, buffer_height) = buffer.size();
+            if buffer_width == width && buffer_height == height {
+                return Ok(());
+            }
+        }
+
+        // Create a new buffer.
+        let buffer = BufferSet::new(&self.display, width, height)?;
+
+        // Set the framebuffer in the CRTC info.
+        // TODO: This requires root access, find a way that doesn't!
```

Comment:
I've had the same issue when implementing KMS support in Slint, and I've found using libseat to be a good solution. When set up, the daemon (for example logind or seatd) return you an FD that's has enough rights to set the frame buffer.

I think that doesn't affect the soft buffer API, but merely your example here (if you wanted to use it).

Reply:
Yes, the `libseat` crate is your best option here to make this work without root wherever possible.